tokio/src/sync/notify.rs

```diff
@@ -516,12 +529,23 @@ impl Notify {
             return;
         }
 
-        // At this point, it is guaranteed that the state will not
-        // concurrently change, as holding the lock is required to
-        // transition **out** of `WAITING`.
+        // Increment the number of times this method was called
+        // and transition to empty.
+        let new_state = set_state(inc_num_notify_waiters_calls(curr), EMPTY);
```

Comment:
Why was this moved up? I'm not sure it matters, but it is not apparent it does *not* matter.

Reply:
The code is correct as long as the store happens while the mutex is held. Moving the store up increases the odds of a `poll_notified` succeeding early in some rare concurrent conditions.

Reply:
It was moved up for correctness. If we allowed to poll a pending future between chunks and observe the old counter value, then it would be possible to observe the inconsistency from the description (number 2.). This is because such future would return `Pending`, even though other waiters from the decoupled list could be already notified. `notify_waiters_poll_consistency_many` checks such scenarios.